apps/cms.c

```diff
@@ -885,10 +892,29 @@ int cms_main(int argc, char **argv)
             goto end;
     }
 
-    in = bio_open_default(infile, 'r',
-                          binary_files ? FORMAT_BINARY : informat);
-    if (in == NULL)
-        goto end;
+    if (digesthex != NULL) {
+        if (infile != NULL || !(flags & CMS_DETACHED)) {
+            BIO_printf(bio_err,
+                       "Cannot use -digest when -in or -nodetach is given\n");
+            goto end;
+        }
+        if (operation != SMIME_SIGN) {
+            BIO_printf(bio_err,
+                       "Warning: -digest is ignored for non-signing operation\n");
```

Comment:
This is IMO incorrect - as `in` needs to be set as in the else clause of the higher level `if` if this is just a warning.

Reply:
I assume this comment refers to the error message "Cannot use -digest when -in or -nodetach option is given" above, which is incorrect. I'm changing it to two separate checks and messages:

* "Cannot use -digest without -in\n" (error)
* "Cannot use -digest when -nodetach is given\n" (error)

Reply:
I'd just turn this warning into an error. IMO it does not make sense for it to be a warning.

Reply:
Sorry, I was too fast.